support-models/src/main/scala/com/gu/support/workers/GiftRecipient.scala

```diff
@@ -54,6 +55,53 @@ object GiftRecipient {
     ))
 }
 
+case class GiftCode private(value: String) extends AnyVal
+
+object GiftCode {
+
+  def apply(value: String): Option[GiftCode] =
+    Some(value)
+      .filter(_.matches(raw"""gd(03|06|12)-[a-km-z02-9]{8}"""))
+      .map(new GiftCode(_))
+
+  implicit val e1 = Encoder.encodeString.contramap[GiftCode](_.value)
+  implicit val d1 = Decoder.decodeString.emap(GiftCode(_).toRight("invalid gift code"))
+
+}
+
+sealed trait GiftRecipientAndMaybeCode {
+  def asDigiSub: Option[DigitalSubGiftRecipientWithCode] = this match { case a: DigitalSubGiftRecipientWithCode => Some(a); case _ => None }
+  def giftRecipient: GiftRecipient
+}
+object GiftRecipientAndMaybeCode {
+
+  case class NonDigitalSubGiftRecipient(giftRecipient: WeeklyGiftRecipient) extends GiftRecipientAndMaybeCode
+  case class DigitalSubGiftRecipientWithCode(
+    giftRecipient: DigitalSubGiftRecipient,
+    giftCode: GiftCode
+  ) extends GiftRecipientAndMaybeCode
```

Comment:
new ADT to hold the gift code - only if it's a DS gift.

I wasn't totally sure about the naming. non digital is basically weekly, but for the purposes of this code we only care that it's not digital.